src/ssh.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "ssh.h"
+
+#include <QProcess>
+
+#include <KConfigGroup>
+#include <KSharedConfig>
+
+#include "settings.h"
+#include "util.h"
+
+QStringList assembleSSHArguments(const QString& deviceName, const QStringList& command)
+{
+    KConfigGroup device = KSharedConfig::openConfig()->group("devices").group(deviceName);
+
+    QStringList arguments = {
+        QStringLiteral("%1@%2").arg(device.readEntry("username", QString()), device.readEntry("hostname", QString()))};
+    const auto sshOptions = device.readEntry("sshoptions", QString());
+    if (!sshOptions.isEmpty()) {
+        arguments.append(sshOptions.split(QLatin1Char(' ')));
+    }
+
+    arguments.append(command);
+
+    return arguments;
+}
+
+QProcessEnvironment sshEnvironment()
+{
+    auto env = Util::appImageEnvironment();
+    Settings* settings = Settings::instance();
+    env.insert(QLatin1String("SSH_ASKPASS"), settings->sshaskpassPath());
+    return env;
+}
+
+QString sshOutput(const QString& deviceName, const QStringList& command)
+{
+    QProcess ssh;
+    ssh.setProgram(QStandardPaths::findExecutable(QLatin1String("ssh")));
+    const auto arguments = assembleSSHArguments(deviceName, command);
+    ssh.setArguments(arguments);
+    ssh.setProcessEnvironment(sshEnvironment());
+    ssh.start();
+    ssh.waitForFinished();
```

Comment:
on a flaky network connection, this will freeze the whole application which is not nice at all. can we find a way to make this async? or at least add timeouts to prevent us from waiting ad infinitum when we drive on a train?

Reply:
I tried making it async, but I didn't find a simple way for this, so I moved the hole thing in another thread (recordpage.cpp:436).